{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## create env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lanhongyi/opt/anaconda3/envs/dsl_project/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.sparse import coo_matrix, save_npz, load_npz\n",
    "from tqdm import tqdm\n",
    "from collections import Counter\n",
    "\n",
    "# constant\n",
    "K_MAX = 6\n",
    "\n",
    "# load data\n",
    "df = pd.read_excel(\"./Data/peptides10.xlsx\")\n",
    "\n",
    "# construct amino to index map\n",
    "amino_acid_index = {\n",
    "    'A': 0,  # Alanine\n",
    "    'R': 1,  # Arginine\n",
    "    'N': 2,  # Asparagine\n",
    "    'D': 3,  # Aspartic acid\n",
    "    'C': 4,  # Cysteine\n",
    "    'E': 5,  # Glutamic acid\n",
    "    'Q': 6,  # Glutamine\n",
    "    'G': 7,  # Glycine\n",
    "    'H': 8,  # Histidine\n",
    "    'I': 9,  # Isoleucine\n",
    "    'L': 10, # Leucine\n",
    "    'K': 11, # Lysine\n",
    "    'M': 12, # Methionine\n",
    "    'F': 13, # Phenylalanine\n",
    "    'P': 14, # Proline\n",
    "    'S': 15, # Serine\n",
    "    'T': 16, # Threonine\n",
    "    'W': 17, # Tryptophan\n",
    "    'Y': 18, # Tyrosine\n",
    "    'V': 19  # Valine\n",
    "}\n",
    "\n",
    "# preprocess data\n",
    "peptides = df['Unnamed: 0']\n",
    "peptides.column = ['Sequence']\n",
    "\n",
    "peptides = peptides.str.replace(' ', '', regex=False)\n",
    "peptides = peptides.str[1:-1]\n",
    "\n",
    "\n",
    "# 存储和读取稀疏矩阵\n",
    "def save_sparse_matrix(mat: pd.DataFrame, path: str) -> None:\n",
    "    # 将pd.Dataframe保存为.npz文件\n",
    "    mat = coo_matrix(mat.values)\n",
    "    save_npz(path, mat)\n",
    "\n",
    "\n",
    "def load_sparse_matrix(path: str) -> pd.DataFrame:\n",
    "    # 读取.npz文件并转换为pd.DataFrame\n",
    "    mat = load_npz(path).toarray()\n",
    "    return pd.DataFrame(mat)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18578</th>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18579</th>\n",
       "      <td>0.250</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18580</th>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18581</th>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18582</th>\n",
       "      <td>0.250</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.125</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18583 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0      1      2      3      4      5      6      7    8      9   \\\n",
       "0      0.125  0.000  0.125  0.000  0.000  0.000  0.000  0.375  0.0  0.000   \n",
       "1      0.000  0.000  0.125  0.125  0.125  0.125  0.000  0.000  0.0  0.125   \n",
       "2      0.000  0.000  0.000  0.250  0.125  0.125  0.000  0.000  0.0  0.125   \n",
       "3      0.000  0.000  0.000  0.250  0.125  0.125  0.000  0.000  0.0  0.125   \n",
       "4      0.125  0.000  0.125  0.000  0.000  0.000  0.000  0.375  0.0  0.000   \n",
       "...      ...    ...    ...    ...    ...    ...    ...    ...  ...    ...   \n",
       "18578  0.125  0.000  0.000  0.000  0.000  0.000  0.125  0.125  0.0  0.125   \n",
       "18579  0.250  0.125  0.000  0.000  0.000  0.000  0.000  0.000  0.0  0.000   \n",
       "18580  0.000  0.125  0.125  0.000  0.000  0.000  0.125  0.125  0.0  0.000   \n",
       "18581  0.125  0.125  0.250  0.000  0.000  0.000  0.125  0.000  0.0  0.125   \n",
       "18582  0.250  0.000  0.000  0.000  0.000  0.125  0.000  0.125  0.0  0.000   \n",
       "\n",
       "          10     11     12     13     14     15     16   17     18     19  \n",
       "0      0.000  0.125  0.125  0.000  0.000  0.125  0.000  0.0  0.000  0.000  \n",
       "1      0.000  0.125  0.000  0.000  0.000  0.000  0.000  0.0  0.250  0.000  \n",
       "2      0.000  0.125  0.000  0.000  0.000  0.000  0.000  0.0  0.250  0.000  \n",
       "3      0.125  0.125  0.000  0.000  0.000  0.000  0.000  0.0  0.125  0.000  \n",
       "4      0.125  0.125  0.000  0.000  0.000  0.125  0.000  0.0  0.000  0.000  \n",
       "...      ...    ...    ...    ...    ...    ...    ...  ...    ...    ...  \n",
       "18578  0.125  0.125  0.000  0.000  0.000  0.000  0.125  0.0  0.000  0.125  \n",
       "18579  0.125  0.000  0.000  0.000  0.125  0.250  0.125  0.0  0.000  0.000  \n",
       "18580  0.125  0.000  0.000  0.000  0.000  0.125  0.000  0.0  0.250  0.000  \n",
       "18581  0.125  0.000  0.000  0.000  0.000  0.000  0.125  0.0  0.000  0.000  \n",
       "18582  0.000  0.000  0.000  0.125  0.000  0.125  0.000  0.0  0.000  0.250  \n",
       "\n",
       "[18583 rows x 20 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_aac_features(peptides: pd.Series) -> pd.DataFrame:\n",
    "    aac_mat = []\n",
    "    for i, peptide in enumerate(peptides):\n",
    "        peptide_aac = [0] * 20\n",
    "        amino_count = Counter(peptide)\n",
    "        for amino in amino_count:\n",
    "            amino_id = amino_acid_index[amino]\n",
    "            peptide_aac[amino_id] = amino_count[amino] / len(peptide)  # 计算每个氨基酸出现的频率\n",
    "        aac_mat.append(peptide_aac)\n",
    "    aac_df = pd.DataFrame(aac_mat)\n",
    "    return aac_df\n",
    "\n",
    "aac_df = extract_aac_features(peptides)\n",
    "save_sparse_matrix(aac_df, './Cache/aac.npz')  # 不保存第一列\n",
    "\n",
    "aac_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PSSM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Align\n",
    "from Bio.Align import substitution_matrices, AlignInfo\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lanhongyi/opt/anaconda3/envs/dsl_project/lib/python3.10/site-packages/Bio/Align/AlignInfo.py:365: BiopythonDeprecationWarning: The `pos_specific_score_matrix` method is deprecated and will be removed in a future release of Biopython. As an alternative, you can convert the multiple sequence alignment object to a new-style Alignment object by via its `.alignment` property, and then create a Motif object. For example, for a multiple sequence alignment `msa` of DNA nucleotides, you would do: \n",
      ">>> alignment = msa.alignment\n",
      ">>> from Bio.motifs import Motif\n",
      ">>> motif = Motif('ACGT', alignment)\n",
      ">>> counts = motif.counts\n",
      "\n",
      "The `counts` object contains the same information as the PSSM returned by `pos_specific_score_matrix`, but note that the indices are reversed:\n",
      "\n",
      ">>> counts[letter][i] == pssm[index][letter]\n",
      "True\n",
      "\n",
      "If your multiple sequence alignment object was obtained using Bio.AlignIO, then you can obtain a new-style Alignment object directly by using Bio.Align.read instead of Bio.AlignIO.read, or Bio.Align.parse instead of Bio.AlignIO.parse.\n",
      "  warnings.warn(\n",
      "/Users/lanhongyi/opt/anaconda3/envs/dsl_project/lib/python3.10/site-packages/Bio/Align/AlignInfo.py:63: BiopythonDeprecationWarning: The `dumb_consensus` method is deprecated and will be removed in a future release of Biopython. As an alternative, you can convert the multiple sequence alignment object to a new-style Alignment object by via its `.alignment` property, and then create a Motif object. You can then use the `.consensus` or `.degenerate_consensus` property of the Motif object to get a consensus sequence. For more control over how the consensus sequence is calculated, you can call the `calculate_consensus` method on the `.counts` property of the Motif object. This is an example for a multiple sequence alignment `msa` of DNA nucleotides:\n",
      ">>> from Bio.Seq import Seq\n",
      ">>> from Bio.SeqRecord import SeqRecord\n",
      ">>> from Bio.Align import MultipleSeqAlignment\n",
      ">>> from Bio.Align.AlignInfo import SummaryInfo\n",
      ">>> msa = MultipleSeqAlignment([SeqRecord(Seq('ACGT')),\n",
      "...                             SeqRecord(Seq('ATGT')),\n",
      "...                             SeqRecord(Seq('ATGT'))])\n",
      ">>> summary = SummaryInfo(msa)\n",
      ">>> dumb_consensus = summary.dumb_consensus(ambiguous='N')\n",
      ">>> print(dumb_consensus)\n",
      "ANGT\n",
      ">>> alignment = msa.alignment\n",
      ">>> from Bio.motifs import Motif\n",
      ">>> motif = Motif('ACGT', alignment)\n",
      ">>> print(motif.consensus)\n",
      "ATGT\n",
      ">>> print(motif.degenerate_consensus)\n",
      "AYGT\n",
      ">>> counts = motif.counts\n",
      ">>> consensus = counts.calculate_consensus(identity=0.7)\n",
      ">>> print(consensus)\n",
      "ANGT\n",
      "\n",
      "If your multiple sequence alignment object was obtained using Bio.AlignIO, then you can obtain a new-style Alignment object directly by using Bio.Align.read instead of Bio.AlignIO.read, or Bio.Align.parse instead of Bio.AlignIO.parse.\n",
      "  warnings.warn(\n",
      "/Users/lanhongyi/opt/anaconda3/envs/dsl_project/lib/python3.10/site-packages/Bio/Align/AlignInfo.py:719: BiopythonDeprecationWarning: The `PSSM` class is deprecated and will be removed in a future release of Biopython. As an alternative, you can convert the multiple sequence alignment object to a new-style Alignment object by via its `.alignment` property, and then create a Motif object. For example, for a multiple sequence alignment `msa` of DNA nucleotides, you would do: \n",
      ">>> alignment = msa.alignment\n",
      ">>> from Bio.motifs import Motif\n",
      ">>> motif = Motif('ACGT', alignment)\n",
      ">>> counts = motif.counts\n",
      "\n",
      "The `counts` object contains the same information as the PSSM returned by `pos_specific_score_matrix`, but note that the indices are reversed:\n",
      "\n",
      ">>> counts[letter][i] == pssm[index][letter]\n",
      "True\n",
      "\n",
      "If your multiple sequence alignment object was obtained using Bio.AlignIO, then you can obtain a new-style Alignment object directly by using Bio.Align.read instead of Bio.AlignIO.read, or Bio.Align.parse instead of Bio.AlignIO.parse.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# 将氨基酸序列列表转换为SeqRecord对象\n",
    "seq_records = [SeqRecord(Seq(seq), id=f\"Seq{i+1}\") for i, seq in enumerate(peptides.values.tolist())]\n",
    "\n",
    "# 创建多序列比对对象\n",
    "alignment = MultipleSeqAlignment(seq_records)\n",
    "\n",
    "# 创建SummaryInfo对象\n",
    "summary = AlignInfo.SummaryInfo(alignment)\n",
    "\n",
    "# 计算PSSM\n",
    "pssm = summary.pos_specific_score_matrix()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A': 1127.0, 'C': 246.0, 'D': 742.0, 'E': 406.0, 'F': 777.0, 'G': 1149.0, 'H': 787.0, 'I': 783.0, 'K': 1019.0, 'L': 954.0, 'M': 871.0, 'N': 757.0, 'P': 811.0, 'Q': 961.0, 'R': 1901.0, 'S': 1702.0, 'T': 1121.0, 'V': 1220.0, 'W': 568.0, 'Y': 681.0}\n",
      "{'A': 1816.0, 'C': 93.0, 'D': 102.0, 'E': 108.0, 'F': 532.0, 'G': 529.0, 'H': 394.0, 'I': 801.0, 'K': 139.0, 'L': 1318.0, 'M': 495.0, 'N': 129.0, 'P': 5486.0, 'Q': 161.0, 'R': 1630.0, 'S': 609.0, 'T': 388.0, 'V': 2494.0, 'W': 637.0, 'Y': 722.0}\n",
      "{'A': 2127.0, 'C': 108.0, 'D': 235.0, 'E': 449.0, 'F': 792.0, 'G': 880.0, 'H': 669.0, 'I': 307.0, 'K': 603.0, 'L': 2054.0, 'M': 895.0, 'N': 680.0, 'P': 207.0, 'Q': 879.0, 'R': 3081.0, 'S': 1776.0, 'T': 369.0, 'V': 668.0, 'W': 948.0, 'Y': 856.0}\n",
      "{'A': 2430.0, 'C': 86.0, 'D': 297.0, 'E': 276.0, 'F': 832.0, 'G': 1521.0, 'H': 1731.0, 'I': 125.0, 'K': 914.0, 'L': 644.0, 'M': 605.0, 'N': 1922.0, 'P': 1024.0, 'Q': 492.0, 'R': 1264.0, 'S': 2239.0, 'T': 293.0, 'V': 195.0, 'W': 919.0, 'Y': 774.0}\n",
      "{'A': 152.0, 'C': 63.0, 'D': 70.0, 'E': 88.0, 'F': 430.0, 'G': 144.0, 'H': 120.0, 'I': 1021.0, 'K': 194.0, 'L': 11122.0, 'M': 1432.0, 'N': 99.0, 'P': 66.0, 'Q': 201.0, 'R': 1661.0, 'S': 200.0, 'T': 82.0, 'V': 454.0, 'W': 407.0, 'Y': 577.0}\n",
      "{'A': 889.0, 'C': 86.0, 'D': 282.0, 'E': 200.0, 'F': 841.0, 'G': 286.0, 'H': 814.0, 'I': 1142.0, 'K': 1236.0, 'L': 1163.0, 'M': 1006.0, 'N': 268.0, 'P': 200.0, 'Q': 1028.0, 'R': 2468.0, 'S': 1581.0, 'T': 2140.0, 'V': 1098.0, 'W': 782.0, 'Y': 1073.0}\n",
      "{'A': 2246.0, 'C': 84.0, 'D': 957.0, 'E': 158.0, 'F': 236.0, 'G': 2069.0, 'H': 167.0, 'I': 283.0, 'K': 1111.0, 'L': 1733.0, 'M': 1126.0, 'N': 1170.0, 'P': 135.0, 'Q': 1040.0, 'R': 1202.0, 'S': 1426.0, 'T': 606.0, 'V': 1601.0, 'W': 174.0, 'Y': 1059.0}\n",
      "{'A': 1221.0, 'C': 81.0, 'D': 185.0, 'E': 1037.0, 'F': 178.0, 'G': 1945.0, 'H': 914.0, 'I': 913.0, 'K': 1041.0, 'L': 1106.0, 'M': 880.0, 'N': 908.0, 'P': 966.0, 'Q': 915.0, 'R': 1134.0, 'S': 1931.0, 'T': 1292.0, 'V': 962.0, 'W': 852.0, 'Y': 122.0}\n"
     ]
    }
   ],
   "source": [
    "for s in pssm:\n",
    "    print(s)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dsl_project",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
